<a href="https://colab.research.google.com/github/LilySu/DS-Unit-1-Sprint-2-Data-Wrangling/blob/master/Lily_Su_Sprint2_DS_Unit_1_Sprint_Challenge_2_Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Data Wrangling

In this Sprint Challenge you will use data from [Gapminder](https://www.gapminder.org/about-gapminder/), a Swedish non-profit co-founded by Hans Rosling. "Gapminder produces free teaching resources making the world understandable based on reliable statistics."
- [Cell phones (total), by country and year](https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--cell_phones_total--by--geo--time.csv)
- [Population (total), by country and year](https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--population_total--by--geo--time.csv)
- [Geo country codes](https://github.com/open-numbers/ddf--gapminder--systema_globalis/blob/master/ddf--entities--geo--country.csv)

These two links have everything you need to successfully complete the Sprint Challenge!
- [Pandas documentation: Working with Text Data](https://pandas.pydata.org/pandas-docs/stable/text.html]) (one question)
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf) (everything else)

## Part 0. Load data

You don't need to add or change anything here. Just run this cell and it loads the data for you, into three dataframes.

In [0]:
import pandas as pd

cell_phones = pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--cell_phones_total--by--geo--time.csv')

population = pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--datapoints--population_total--by--geo--time.csv')

geo_country_codes = (pd.read_csv('https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/master/ddf--entities--geo--country.csv')
                       .rename(columns={'country': 'geo', 'name': 'country'}))

## Part 1. Join data

First, join the `cell_phones` and `population` dataframes (with an inner join on `geo` and `time`).

The resulting dataframe's shape should be: (8590, 4)

In [2]:
cell_phones.shape, population.shape, geo_country_codes.shape

((9215, 3), (59297, 3), (273, 33))

In [3]:
cell_phones.head(3)

geo  time  cell_phones_total
0  abw  1960                0.0
1  abw  1965                0.0
2  abw  1970                0.0

In [4]:
population.head(3)

geo  time  population_total
0  afg  1800           3280000
1  afg  1801           3280000
2  afg  1802           3280000

In [5]:
cell_phones_by_pop = pd.merge(cell_phones, population, how='outer', on=['geo', 'time'])
cell_phones_by_pop.head(3)

geo  time  cell_phones_total  population_total
0  abw  1960                0.0               NaN
1  abw  1965                0.0               NaN
2  abw  1970                0.0               NaN

In [6]:
cell_phones_by_pop.shape

(59922, 4)

In [77]:
cell_phones_by_pop_nona = cell_phones_by_pop.dropna()
cell_phones_by_pop_nona.shape

(8590, 4)

Then, select the `geo` and `country` columns from the `geo_country_codes` dataframe, and join with your population and cell phone data.

The resulting dataframe's shape should be: (8590, 5)

In [8]:
geo_country_codes.head(3)

geo alt_5                    alternative_1 alternative_2 alternative_3  \
0  abkh   NaN                              NaN           NaN           NaN   
1   abw   NaN                              NaN           NaN           NaN   
2   afg   NaN  Islamic Republic of Afghanistan           NaN           NaN   

  alternative_4_cdiac arb1 arb2 arb3 arb4         ...          latitude  \
0                 NaN  NaN  NaN  NaN  NaN         ...               NaN   
1               Aruba  NaN  NaN  NaN  NaN         ...              12.5   
2         Afghanistan  NaN  NaN  NaN  NaN         ...              33.0   

  longitude main_religion_2008      country        pandg un_state  \
0       NaN                NaN     Abkhazia          NaN    False   
1 -69.96667          christian        Aruba          NaN    False   
2  66.00000             muslim  Afghanistan  AFGHANISTAN     True   

  unicode_region_subtag  upper_case_name world_4region        world_6region  
0                   NaN              NaN        europe  europe_central_asia  
1                    AW            ARUBA      americas              america  
2                    AF      AFGHANISTAN          asia           south_asia  

[3 rows x 33 columns]

In [9]:
geo_country_codes.T

0            1    \
geo                                    abkh          abw   
alt_5                                   NaN          NaN   
alternative_1                           NaN          NaN   
alternative_2                           NaN          NaN   
alternative_3                           NaN          NaN   
alternative_4_cdiac                     NaN        Aruba   
arb1                                    NaN          NaN   
arb2                                    NaN          NaN   
arb3                                    NaN          NaN   
arb4                                    NaN          NaN   
arb5                                    NaN          NaN   
arb6                                    NaN          NaN   
g77_and_oecd_countries               others       others   
gapminder_list                     Abkhazia        Aruba   
god_id                                GE-AB           AW   
gwid                                     i0          i12   
income_groups                           NaN  high_income   
is--country                            True         True   
iso3166_1_alpha2                        NaN           AW   
iso3166_1_alpha3                        NaN          ABW   
iso3166_1_numeric                       NaN          533   
iso3166_2                               NaN          NaN   
landlocked                              NaN    coastline   
latitude                                NaN         12.5   
longitude                               NaN     -69.9667   
main_religion_2008                      NaN    christian   
country                            Abkhazia        Aruba   
pandg                                   NaN          NaN   
un_state                              False        False   
unicode_region_subtag                   NaN           AW   
upper_case_name                         NaN        ARUBA   
world_4region                        europe     americas   
world_6region           europe_central_asia      america   

                                                    2                    3    \
geo                                                 afg                  ago   
alt_5                                               NaN                  NaN   
alternative_1           Islamic Republic of Afghanistan                  NaN   
alternative_2                                       NaN                  NaN   
alternative_3                                       NaN                  NaN   
alternative_4_cdiac                         Afghanistan               Angola   
arb1                                                NaN                  NaN   
arb2                                                NaN                  NaN   
arb3                                                NaN                  NaN   
arb4                                                NaN                  NaN   
arb5                                                NaN                  NaN   
arb6                                                NaN                  NaN   
g77_and_oecd_countries                              g77                  g77   
gapminder_list                              Afghanistan               Angola   
god_id                                               AF                   AO   
gwid                                                 i1                   i7   
income_groups                                low_income  upper_middle_income   
is--country                                        True                 True   
iso3166_1_alpha2                                     AF                   AO   
iso3166_1_alpha3                                    AFG                  AGO   
iso3166_1_numeric                                     4                   24   
iso3166_2                                           NaN                  NaN   
landlocked                                   landlocked            coastline   
latitude                                             33                -12.5   

In [140]:
geo_country_codes_just_geo_country = geo_country_codes.iloc[:,[0, 26]]
geo_country_codes_just_geo_country.head(3)


geo      country
0  abkh     Abkhazia
1   abw        Aruba
2   afg  Afghanistan

In [99]:
geo_country_codes_just_geo_country.shape, cell_phones_by_pop_nona.shape

((273, 2), (8590, 4))

In [100]:
cell_phones_by_pop_nona.head(3)

geo  time  cell_phones_total  population_total
41  afg  1960                0.0       8,996,351.0
42  afg  1965                0.0       9,938,414.0
43  afg  1970                0.0      11,126,123.0

In [141]:
cell_phones_by_pop_updated = pd.merge(cell_phones_by_pop_nona, geo_country_codes_just_geo_country,how='left', on='geo')
cell_phones_by_pop_updated.shape

(8590, 5)

## Part 2. Make features

Calculate the number of cell phones per person, and add this column onto your dataframe.

(You've calculated correctly if you get 1.220 cell phones per person in the United States in 2017.)

In [29]:
cell_phones_by_pop_updated['cell_phones_total'].nunique()

4491

In [31]:
(cell_phones_by_pop_updated.groupby('cell_phones_total')
              .apply(lambda x: x.groupby('geo').first().sum())
              .reset_index(name='unique_users'))

TypeError: ignored

In [34]:
# cell_phones_by_pop_updated.drop_duplicates(['geo']).groupby('country').agg({'cell_phones_total':'sum'})

cell_phones_total
country                                  
Afghanistan                             0
Albania                                 0
Algeria                                 0
Andorra                                 0
Angola                                  0
Antigua and Barbuda                     0
Argentina                               0
Armenia                                 0
Australia                               0
Austria                                 0
Azerbaijan                              0
Bahamas                                 0
Bahrain                                 0
Bangladesh                              0
Barbados                                0
Belarus                                 0
Belgium                                 0
Belize                                  0
Benin                                   0
Bhutan                                  0
Bolivia                                 0
Bosnia and Herzegovina                  0
Botswana                                0
Brazil                                  0
Brunei                                  0
Bulgaria                                0
Burkina Faso                            0
Burundi                                 0
Cambodia                                0
Cameroon                                0
...                                   ...
Sudan                                   0
Suriname                                0
Swaziland                               0
Sweden                                  0
Switzerland                             0
Syria                                   0
Tajikistan                              0
Tanzania                                0
Thailand                                0
Timor-Leste                             0
Togo                                    0
Tonga                                   0
Trinidad and Tobago                     0
Tunisia                                 0
Turkey                                  0
Turkmenistan                            0
Tuvalu                                  0
Uganda                                  0
Ukraine                                 0
United Arab Emirates                 2330
United Kingdom                          0
United States                           0
Uruguay                                 0
Uzbekistan                              0
Vanuatu                                 0
Venezuela                               0
Vietnam                                 0
Yemen                                   0
Zambia                                  0
Zimbabwe                                0

[195 rows x 1 columns]

In [104]:
# aggregation_functions = {'cell_phones_total': 'sum'}
# cell_phones_by_pop_updated_new = cell_phones_by_pop_updated.groupby(cell_phones_by_pop_updated['country']).aggregate(aggregation_functions)
# cell_phones_by_pop_updated_new.head(3)

cell_phones_total
country                       
Afghanistan      167,422,262.0
Albania           40,347,847.0
Algeria          457,970,229.0

In [105]:
cell_phones_by_pop_updated_new.dtypes

cell_phones_total    float64
dtype: object

In [106]:
cell_phones_by_pop_updated.head(3)

geo  time  cell_phones_total  population_total      country
0  afg  1960                0.0       8,996,351.0  Afghanistan
1  afg  1965                0.0       9,938,414.0  Afghanistan
2  afg  1970                0.0      11,126,123.0  Afghanistan

In [107]:
cell_phones_by_pop_updated.dtypes

geo                   object
time                   int64
cell_phones_total    float64
population_total     float64
country               object
dtype: object

In [0]:
cell_phones_by_pop_updated['population_total']=cell_phones_by_pop_updated['population_total'].astype(float)

In [109]:
cell_phones_by_pop_updated.dtypes

geo                   object
time                   int64
cell_phones_total    float64
population_total       int64
country               object
dtype: object

In [0]:
# cell_phones_by_pop_updated_condensed = cell_phones_by_pop_updated.groupby(['country']).mean()

In [142]:
cell_phones_by_pop_updated.head(3)

geo  time  cell_phones_total  population_total      country
0  afg  1960                0.0       8,996,351.0  Afghanistan
1  afg  1965                0.0       9,938,414.0  Afghanistan
2  afg  1970                0.0      11,126,123.0  Afghanistan

In [0]:
pd.set_option('display.max_rows', 1000)

In [161]:
# cell_phones_by_pop_updated['mean'] = cell_phones_by_pop_updated['population_total']/cell_phones_by_pop_updated['cell_phones_total']
pd.set_option('display.max_rows', 1000)
cell_phones_by_pop_updated['mean'] = cell_phones_by_pop_updated['population_total']/cell_phones_by_pop_updated['cell_phones_total']
cell_phones_by_pop_updated

geo    time  cell_phones_total  population_total  \
0     AFG 1,960.0                0.0       8,996,351.0   
1     AFG 1,965.0                0.0       9,938,414.0   
2     AFG 1,970.0                0.0      11,126,123.0   
3     AFG 1,975.0                0.0      12,590,286.0   
4     AFG 1,976.0                0.0      12,840,299.0   
5     AFG 1,977.0                0.0      13,067,538.0   
6     AFG 1,978.0                0.0      13,237,734.0   
7     AFG 1,979.0                0.0      13,306,695.0   
8     AFG 1,980.0                0.0      13,248,370.0   
9     AFG 1,981.0                0.0      13,053,954.0   
10    AFG 1,982.0                0.0      12,749,645.0   
11    AFG 1,983.0                0.0      12,389,269.0   
12    AFG 1,984.0                0.0      12,047,115.0   
13    AFG 1,985.0                0.0      11,783,050.0   
14    AFG 1,986.0                0.0      11,601,041.0   
15    AFG 1,987.0                0.0      11,502,761.0   
16    AFG 1,988.0                0.0      11,540,888.0   
17    AFG 1,989.0                0.0      11,777,609.0   
18    AFG 1,990.0                0.0      12,249,114.0   
19    AFG 1,991.0                0.0      12,993,657.0   
20    AFG 1,992.0                0.0      13,981,231.0   
21    AFG 1,993.0                0.0      15,095,099.0   
22    AFG 1,994.0                0.0      16,172,719.0   
23    AFG 1,995.0                0.0      17,099,541.0   
24    AFG 1,996.0                0.0      17,822,884.0   
25    AFG 1,997.0                0.0      18,381,605.0   
26    AFG 1,998.0                0.0      18,863,999.0   
27    AFG 1,999.0                0.0      19,403,676.0   
28    AFG 2,000.0                0.0      20,093,756.0   
29    AFG 2,001.0                0.0      20,966,463.0   
30    AFG 2,002.0           25,000.0      21,979,923.0   
31    AFG 2,003.0          200,000.0      23,064,851.0   
32    AFG 2,004.0          600,000.0      24,118,979.0   
33    AFG 2,005.0        1,200,000.0      25,070,798.0   
34    AFG 2,006.0        2,520,366.0      25,893,450.0   
35    AFG 2,007.0        4,668,096.0      26,616,792.0   
36    AFG 2,008.0        7,898,909.0      27,294,031.0   
37    AFG 2,009.0       10,500,000.0      28,004,331.0   
38    AFG 2,010.0       10,215,840.0      28,803,167.0   
39    AFG 2,011.0       13,797,879.0      29,708,599.0   
40    AFG 2,012.0       15,340,115.0      30,696,958.0   
41    AFG 2,013.0       16,807,156.0      31,731,688.0   
42    AFG 2,014.0       18,407,168.0      32,758,020.0   
43    AFG 2,015.0       19,709,038.0      33,736,494.0   
44    AFG 2,016.0       21,602,982.0      34,656,032.0   
45    AFG 2,017.0       23,929,713.0      35,530,081.0   
46    AGO 1,960.0                0.0       5,643,182.0   
47    AGO 1,965.0                0.0       6,203,299.0   
48    AGO 1,970.0                0.0       6,776,381.0   
49    AGO 1,975.0                0.0       7,682,479.0   
50    AGO 1,976.0                0.0       7,900,997.0   
51    AGO 1,977.0                0.0       8,130,988.0   
52    AGO 1,978.0                0.0       8,376,147.0   
53    AGO 1,979.0                0.0       8,641,521.0   
54    AGO 1,980.0                0.0       8,929,900.0   
55    AGO 1,981.0                0.0       9,244,507.0   
56    AGO 1,982.0                0.0       9,582,156.0   
57    AGO 1,983.0                0.0       9,931,562.0   
58    AGO 1,984.0                0.0      10,277,321.0   
59    AGO 1,985.0                0.0      10,609,042.0   
60    AGO 1,986.0                0.0      10,921,037.0   
61    AGO 1,987.0                0.0      11,218,268.0   
62    AGO 1,988.0                0.0      11,513,968.0   
63    AGO 1,989.0                0.0      11,827,237.0   
64    AGO 1,990.0                0.0      12,171,441.0   
65    AGO 1,991.0                0.0      12,553,446.0   
66    AGO 1,992.0                0.0      12,968,345.0   
67    AGO 1,993.0            1,100.0      13,403,734.0   
68  

Modify the `geo` column to make the geo codes uppercase instead of lowercase.

In [66]:
# cell_phones_by_pop_nona.dtypes

geo                  object
time                   |S21
cell_phones_total      |S32
population_total       |S32
dtype: object

In [63]:
# cell_phones_by_pop_nona['geo'] = cell_phones_by_pop_nona.loc[:0].astype('|S')

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:537: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [156]:
cell_phones_by_pop_updated['geo'] = cell_phones_by_pop_updated['geo'].str.upper()
# cell_phones_by_pop_nona['geo'] = cell_phones_by_pop_nona['geo'].str.upper()
# cell_phones_by_pop_nona.head(3)
cell_phones_by_pop_updated.head(3)

geo    time  cell_phones_total  population_total      country
0  AFG 1,960.0                0.0       8,996,351.0  Afghanistan
1  AFG 1,965.0                0.0       9,938,414.0  Afghanistan
2  AFG 1,970.0                0.0      11,126,123.0  Afghanistan

In [154]:
# cell_phones_by_pop_nona.dtypes

geo                   object
time                   int64
cell_phones_total    float64
population_total     float64
dtype: object

## Part 3. Process data

Use the describe function, to describe your dataframe's numeric columns, and then its non-numeric columns.

(You'll see the time period ranges from 1960 to 2017, and there are 195 unique countries represented.)

In [162]:
cell_phones_by_pop_updated.describe()

time    cell_phones_total     population_total  \
count             8,590.0              8,590.0              8,590.0   
mean  1,994.1934807916182  9,004,949.642905472 29,838,230.581722934   
std    14.257974607310302 55,734,084.872179635 116,128,377.47477299   
min               1,960.0                  0.0              4,433.0   
25%               1,983.0                  0.0          1,456,148.0   
50%               1,995.0              6,200.0          5,725,062.5   
75%               2,006.0          1,697,652.0         18,105,812.0   
max               2,017.0      1,474,097,000.0      1,409,517,397.0   

                    mean  
count            8,590.0  
mean                 inf  
std                  nan  
min   0.4015672658756129  
25%   2.1684958046688436  
50%    639.5388643931718  
75%                  inf  
max                  inf

In [172]:
cell_phones_by_pop_updated.head(3)

geo    time  cell_phones_total  population_total      country  mean
0  AFG 1,960.0                0.0       8,996,351.0  Afghanistan   inf
1  AFG 1,965.0                0.0       9,938,414.0  Afghanistan   inf
2  AFG 1,970.0                0.0      11,126,123.0  Afghanistan   inf

In 2017, what were the top 5 countries with the most cell phones total?

Your list of countries should have these totals:

| country | cell phones total |
|:-------:|:-----------------:|
|    ?    |     1,474,097,000 |
|    ?    |     1,168,902,277 |
|    ?    |       458,923,202 |
|    ?    |       395,881,000 |
|    ?    |       236,488,548 |



In [175]:
just_2017 = cell_phones_by_pop_updated.loc[cell_phones_by_pop_updated['country'] == '2017']
just_2017.head(5)

Empty DataFrame
Columns: [geo, time, cell_phones_total, population_total, country, mean]
Index: []

In [167]:
cell_phones_by_pop_updated['cell_phones_total'].sort_values().head(5)


0      0.0
4308   0.0
4307   0.0
4306   0.0
8092   0.0
Name: cell_phones_total, dtype: float64

In [0]:
# This optional code formats float numbers with comma separators
pd.options.display.float_format = '{:,}'.format

2017 was the first year that China had more cell phones than people.

What was the first year that the USA had more cell phones than people?

In [0]:
def phones_than_people(x):
    

## Part 4. Reshape data

Create a pivot table:
- Columns: Years 2007—2017
- Rows: China, India, United States, Indonesia, Brazil (order doesn't matter)
- Values: Cell Phones Total

The table's shape should be: (5, 11)

In [179]:
cell_phones_by_pop_updated.head(3)

time         1,960.0  1,965.0  1,970.0  1,975.0  1,976.0  1,977.0  1,978.0  \
country                                                                      
Afghanistan      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
Albania          0.0      0.0      0.0      0.0      0.0      0.0      0.0   
Algeria          0.0      0.0      0.0      0.0      0.0      0.0      0.0   

time         1,979.0  1,980.0  1,981.0     ...           2,008.0      2,009.0  \
country                                    ...                                  
Afghanistan      0.0      0.0      0.0     ...       7,898,909.0 10,500,000.0   
Albania          0.0      0.0      0.0     ...       1,859,632.0  2,463,741.0   
Algeria          0.0      0.0      0.0     ...      27,031,472.0 32,729,824.0   

time             2,010.0      2,011.0      2,012.0      2,013.0      2,014.0  \
country                                                                        
Afghanistan 10,215,840.0 13,797,879.0 15,340,115.0 16,807,156.0 18,407,168.0   
Albania      2,692,372.0  3,100,000.0  3,500,000.0  3,685,983.0  3,359,654.0   
Algeria     32,780,165.0 35,615,926.0 37,527,703.0 39,517,045.0 43,298,174.0   

time             2,015.0      2,016.0      2,017.0  
country                                             
Afghanistan 19,709,038.0 21,602,982.0 23,929,713.0  
Albania      3,400,955.0  3,369,756.0  3,497,950.0  
Algeria     43,227,643.0 47,041,321.0 49,873,389.0  

[3 rows x 46 columns]

In [178]:
cell_phones_by_pop_updated = cell_phones_by_pop_updated.pivot_table(index='country',columns='time',values='cell_phones_total')
cell_phones_by_pop_updated.head(5)

KeyError: ignored

#### OPTIONAL BONUS QUESTION!

Sort these 5 countries, by biggest increase in cell phones from 2007 to 2017.

Which country had 935,282,277 more cell phones in 2017 versus 2007?

If you have the time and curiosity, what other questions can you ask and answer with this data?